e2e/framework/nacos.go

```diff
@@ -0,0 +1,202 @@
+package framework
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	"github.com/nacos-group/nacos-sdk-go/clients"
+	"github.com/nacos-group/nacos-sdk-go/clients/config_client"
+	"github.com/nacos-group/nacos-sdk-go/common/constant"
+	"github.com/nacos-group/nacos-sdk-go/vo"
+	"github.com/onsi/ginkgo"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	e2econst "github.com/api7/apisix-mesh-agent/e2e/framework/constant"
+	"github.com/api7/apisix-mesh-agent/pkg/log"
+)
+
+func nacosNamespace() string {
+	return "nacos-e2e"
+}
+
+type NacosConfig struct {
+	DataId  string
+	Content string
+}
+
+type NacosInstallation struct {
+	Preinstalled bool
+	KubectlOpts  *k8s.KubectlOptions
+
+	Namespace   string
+	ChartPath   string
+	ReleaseName string
+
+	Tunnel       *k8s.Tunnel
+	Endpoint     string
+	ConfigClient config_client.IConfigClient
+
+	mu   sync.Mutex
+	cfgs map[string]*NacosConfig
+}
+
+// NewNacos installs nacos
+func NewNacos() (*NacosInstallation, error) {
+	nacos := &NacosInstallation{
+		KubectlOpts: &k8s.KubectlOptions{
+			ConfigPath: GetKubeconfig(),
+			Namespace:  nacosNamespace(),
+		},
+		Namespace:   nacosNamespace(),
+		ChartPath:   filepath.Join(e2econst.E2eHome, "charts/nacos"),
+		ReleaseName: "nacos-e2e",
+		cfgs:        make(map[string]*NacosConfig),
+	}
+
+	err := k8s.CreateNamespaceE(ginkgo.GinkgoT(), nacos.KubectlOpts, nacos.Namespace)
+	if err != nil {
+		if !strings.Contains(err.Error(), "already exists") {
+			log.Errorw("failed to create namespace nacos",
+				zap.Error(err),
+			)
+			return nil, err
+		}
+	}
+
+	_, err = k8s.GetPodE(ginkgo.GinkgoT(), nacos.KubectlOpts, nacos.ReleaseName+"-0")
+	if err != nil {
+		if !strings.Contains(err.Error(), "not found") {
+			return nil, err
+		}
+
+		nacos.Preinstalled = false
+		// install
+		install := exec.Command(e2econst.Helm, "install", nacos.ReleaseName,
+			"--namespace", nacos.Namespace, "--kubeconfig", GetKubeconfig(), nacos.ChartPath)
+		installErr := bytes.NewBuffer(nil)
+		install.Stderr = installErr
+		err = install.Run()
+		if err != nil {
+			log.Errorw("failed to run nacos install command",
+				zap.String("command", install.String()),
+				zap.Error(err),
+				zap.String("stderr", installErr.String()),
+			)
+			return nil, err
+		}
+	} else {
+		nacos.Preinstalled = true
+	}
+
+	condFunc := func() (bool, error) {
+		pod, err := k8s.GetPodE(ginkgo.GinkgoT(), nacos.KubectlOpts, nacos.ReleaseName+"-0")
+		if err != nil {
+			return false, err
```

Comment:
Should not return `err` directly if you want to retry as `wait.PollImmediate` will return immediately if `err` is not `nil`.